From b23f4973e3747e00b2a8fa63a45d78142f1c7c2f Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@kde.org>
Date: Fri, 28 Aug 2026 13:19:08 +0200
Subject: [PATCH 1/3] backends/drm: fix a color profile source check

It always checked the SDR source, rather than the correct one for the
current mode.

CCBUG: 524038
---
 src/backends/drm/drm_output.cpp | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/backends/drm/drm_output.cpp b/src/backends/drm/drm_output.cpp
index 189dd5727de..1561bcbeed2 100644
--- a/src/backends/drm/drm_output.cpp
+++ b/src/backends/drm/drm_output.cpp
@@ -615,12 +615,12 @@ bool DrmOutput::queueChanges(const std::shared_ptr<OutputChangeSet> &props)
         m_pipeline->setMaxBpc(decideAutomaticBpcLimit().value_or(tradeoff == ColorPowerTradeoff::PreferAccuracy ? 16 : 10));
     }
 
-    if (m_nextState->colorProfileSource != ColorProfileSource::ICC || (bt2020 && !hdr)) {
-        m_pipeline->setIccProfile(nullptr);
-    } else if (hdr) {
+    if (hdr && m_nextState->hdrColorProfileSource == ColorProfileSource::ICC) {
         m_pipeline->setIccProfile(m_nextState->hdrIccProfile);
-    } else {
+    } else if (!hdr && !bt2020 && m_nextState->colorProfileSource == ColorProfileSource::ICC) {
         m_pipeline->setIccProfile(m_nextState->iccProfile);
+    } else {
+        m_pipeline->setIccProfile(nullptr);
     }
     // remove the color pipeline for the atomic test
     // otherwise it could potentially fail
-- 
GitLab


From eb6a57fba628f700e2046dafc3367677ee29ce6a Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@kde.org>
Date: Fri, 28 Aug 2026 13:24:36 +0200
Subject: [PATCH 2/3] core/iccprofile: fall back to BToA0 if BToA1 isn't
 available

The spec says to fall back to BToA0 if the tag for the used rendering
intent can't be used, and to fall back to matrix+shaper if that can't
be used either (which we do already).

CCBUG: 524038
---
 src/core/iccprofile.cpp | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/core/iccprofile.cpp b/src/core/iccprofile.cpp
index 6a2e184e308..937b791c2c2 100644
--- a/src/core/iccprofile.cpp
+++ b/src/core/iccprofile.cpp
@@ -106,7 +106,13 @@ const ColorPipeline *IccProfile::BToATag(RenderingIntent intent) const
         // but that has to be handled before the tag is applied
     case RenderingIntent::RelativeColorimetricWithBPC:
     case RenderingIntent::AbsoluteColorimetricNoAdaptation:
-        return m_bToA1Tag ? &*m_bToA1Tag : nullptr;
+        if (m_bToA1Tag) {
+            return &*m_bToA1Tag;
+        } else if (m_bToA0Tag) {
+            return &*m_bToA0Tag;
+        } else {
+            return nullptr;
+        }
     }
     Q_UNREACHABLE();
 }
-- 
GitLab


From 59ae700f62d68483f6154354fc3661437c0bb468 Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@kde.org>
Date: Fri, 28 Aug 2026 13:37:24 +0200
Subject: [PATCH 3/3] opengl/icc_shader: handle multipliers, too

ColorPipeline automatically optimizes scaling-only matrices to a
multiplier, which is normally very helpful, but in this case caused
the shader code to break.

BUG: 524038
---
 src/opengl/icc_shader.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/opengl/icc_shader.cpp b/src/opengl/icc_shader.cpp
index 79f93ed2698..7ef748f2422 100644
--- a/src/opengl/icc_shader.cpp
+++ b/src/opengl/icc_shader.cpp
@@ -108,6 +108,10 @@ bool IccShader::setProfile(const std::shared_ptr<IccProfile> &profile, const std
             if (it != tag->ops.end() && std::holds_alternative<ColorMatrix>(it->operation)) {
                 matrix2 = std::get<ColorMatrix>(it->operation).mat;
                 it++;
+            } else if (it != tag->ops.end() && std::holds_alternative<ColorMultiplier>(it->operation)) {
+                const auto mult = std::get<ColorMultiplier>(it->operation).factors;
+                matrix2.scale(mult);
+                it++;
             }
             if (it != tag->ops.end() && std::holds_alternative<std::shared_ptr<ColorTransformation>>(it->operation)) {
                 const auto sample = [&op = std::get<std::shared_ptr<ColorTransformation>>(it->operation)](size_t x) {
-- 
GitLab

